metal/eip_controlplane_reconciliation.go

```diff
@@ -69,138 +74,156 @@ func newControlPlaneEndpointManager(k8sclient kubernetes.Interface, stop <-chan
 			}},
 		eipTag:        eipTag,
 		projectID:     projectID,
-		instances:     i,
 		ipResSvr:      ipResSvr,
 		deviceIPSrv:   deviceIPSrv,
 		apiServerPort: apiServerPort,
 		k8sclient:     k8sclient,
-		stop:          stop,
 	}
-	klog.V(2).Info("controlPlaneEndpointManager.init(): enabling BGP on project")
+
 	ctx, cancel := context.WithCancel(context.Background())
 	go func() {
 		<-stop
 		cancel()
 	}()
 
-	sharedInformer := informers.NewSharedInformerFactory(k8sclient, checkLoopTimerSeconds*time.Second)
-	var watchers []cache.SharedIndexInformer
-	var name = "controlplaneEndpointManager"
-
-	// register to capture all new services
-	servicesInformer := sharedInformer.Core().V1().Services().Informer()
-	servicesInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			svc := obj.(*v1.Service)
-			if err := m.reconcileServices(ctx, []*v1.Service{svc}, ModeAdd); err != nil {
-				klog.Errorf("%s failed to update and sync service for add %s/%s: %v", name, svc.Namespace, svc.Name, err)
-			}
-		},
-		DeleteFunc: func(obj interface{}) {
-			svc := obj.(*v1.Service)
-			if err := m.reconcileServices(ctx, []*v1.Service{svc}, ModeRemove); err != nil {
-				klog.Errorf("%s failed to update and sync service for remove %s/%s: %v", name, svc.Namespace, svc.Name, err)
-			}
-		},
-	})
+	for _, label := range controlPlaneLabels {
+		req, err := labels.NewRequirement(label, selection.Exists, nil)
+		if err != nil {
+			return m, err
+		}
 
-	if servicesInformer != nil {
-		watchers = append(watchers, servicesInformer)
+		m.controlPlaneSelectors = append(m.controlPlaneSelectors, labels.NewSelector().Add(*req))
 	}
 
-	nodesInformer := sharedInformer.Core().V1().Nodes().Informer()
-	nodesInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			n := obj.(*v1.Node)
-			if err := m.reconcileNodes(ctx, []*v1.Node{n}, ModeAdd); err != nil {
-				klog.Errorf("%s failed to update and sync node for add %s for handler: %v", name, n.Name, err)
-			}
-		},
-		DeleteFunc: func(obj interface{}) {
-			n := obj.(*v1.Node)
-			if err := m.reconcileNodes(ctx, []*v1.Node{n}, ModeRemove); err != nil {
-				klog.Errorf("%s failed to update and sync node for remove %s for handler: %v", name, n.Name, err)
-			}
+	sharedInformer := informers.NewSharedInformerFactory(k8sclient, checkLoopTimerSeconds*time.Second)
+
+	// register to capture all new nodes
+	sharedInformer.Core().V1().Nodes().Informer().AddEventHandler(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				n, _ := obj.(*v1.Node)
+				return m.nodeFilter(n)
+			},
```

Comment:
Given that the `apiServerPort` might not be set the first time a node is added, will this work the next time? Is the `FilterFunc` called on every sync?

Reply:
FilterFunc is called prior to calling any of the Handler functions. So it'll be called for any watch based events or resync events.

Reply:
Since we were attempting to default these values in both the node filter and during Services reconciliation, I went ahead and deduplicated it to only be done during the Services reconciliation instead.